File: backend/evm/src/main/java/com/earnedvaluemanagement/evm/service/ActivityService.java
package com.earnedvaluemanagement.evm.service;

import com.earnedvaluemanagement.evm.dto.request.ActivityRequest;
import com.earnedvaluemanagement.evm.dto.response.ActivityResponse;

import java.util.List;

public interface ActivityService {

    List<ActivityResponse> findAllByProjectId(Long projectId);

    ActivityResponse findById(Long projectId, Long activityId);

    ActivityResponse create(Long projectId, ActivityRequest request);

    ActivityResponse update(Long projectId, Long activityId, ActivityRequest request);

    void delete(Long projectId, Long activityId);
}

File: frontend/earned-value-management-frontend/src/app/core/services/activity.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { lastValueFrom } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Activity, ActivityRequest } from '../models/activity.model';

@Injectable({ providedIn: 'root' })
export class ActivityService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = environment.apiUrl;

  private url(projectId: number): string {
    return `${this.baseUrl}/projects/${projectId}/activities`;
  }

  findAll(projectId: number): Promise<Activity[]> {
    return lastValueFrom(this.http.get<Activity[]>(this.url(projectId)));
  }

  findById(projectId: number, activityId: number): Promise<Activity> {
    return lastValueFrom(this.http.get<Activity>(`${this.url(projectId)}/${activityId}`));
  }

  create(projectId: number, request: ActivityRequest): Promise<Activity> {
    return lastValueFrom(this.http.post<Activity>(this.url(projectId), request));
  }

  update(projectId: number, activityId: number, request: ActivityRequest): Promise<Activity> {
    return lastValueFrom(this.http.put<Activity>(`${this.url(projectId)}/${activityId}`, request));
  }

  delete(projectId: number, activityId: number): Promise<void> {
    return lastValueFrom(this.http.delete<void>(`${this.url(projectId)}/${activityId}`));
  }
}
